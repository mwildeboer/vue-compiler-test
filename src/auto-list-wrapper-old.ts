import { parse, walk, ELEMENT_NODE, render } from "ultrahtml";
import type { ElementNode, Node } from "ultrahtml";

const V_FOR_DIRECTIVE = "v-for";
const V_FOR_KEY_DIRECTIVE = ":key";
const V_IF_DIRECTIVE = "v-if";
const V_ELSE_IF_DIRECTIVE = "v-else-if";
const V_ELSE_DIRECTIVE = "v-else";

function transferDirectiveToWrapper(node: Node, wrapper: Node, directive: string) {
  if (directive in node.attributes) {
    wrapper.attributes[directive] = node.attributes[directive];

    delete node.attributes[directive];
  }
}

export async function autoLitWrapperOld(template: string, elementName: string, filter: (node: ElementNode) => boolean) {
  const ast = parse(template);

  await walk(ast, (node) => {
    if (node?.attributes) {
      delete node.attributes[""];
    }

    if (node.type !== ELEMENT_NODE || !filter(node)) {
      return;
    }

    const wrapper: Node = {
      name: elementName,
      type: ELEMENT_NODE,
      parent: node.parent,
      children: [node],
      attributes: {},
      loc: node.loc
    };

    if (node.attributes[V_FOR_DIRECTIVE]) {
      transferDirectiveToWrapper(node, wrapper, V_FOR_DIRECTIVE);
      transferDirectiveToWrapper(node, wrapper, V_FOR_KEY_DIRECTIVE);
    }

    transferDirectiveToWrapper(node, wrapper, V_IF_DIRECTIVE);
    transferDirectiveToWrapper(node, wrapper, V_ELSE_IF_DIRECTIVE);
    transferDirectiveToWrapper(node, wrapper, V_ELSE_DIRECTIVE);

    delete node.attributes[""];

    node.parent.children.splice(node.parent.children.indexOf(node), 1, wrapper);
  });

  const transformedCode = await render(ast);
  return transformedCode;
}
