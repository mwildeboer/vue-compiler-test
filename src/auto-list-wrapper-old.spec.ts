import { expect, test, describe } from "vitest";
import { autoLitWrapperOld } from "./auto-list-wrapper-old";
import { ElementNode } from "ultrahtml";

const filter = (node: ElementNode) => ("name" in node ? node.name.startsWith("kpn-") : false);

describe("autoLitWrapper", () => {
  test("simple", async () => {
    const template = `<kpn-title id="title" :foo="bar">Hello world!</kpn-title>`;
    const string = await autoLitWrapperOld(template, "LitWrapper", filter);
    expect(string).toMatchSnapshot();
  });

  test("complex", async () => {
    const template = `
    <kpn-title v-if="true" id="title" :foo="bar > 1 ? 2 : 1" bind:test="test">
        <span>Hello world!</span>
        <img src="src" />
    </kpn-title>
    `;

    const string = await autoLitWrapperOld(template, "LitWrapper", filter);
    expect(string).toMatchSnapshot();
  });

  test("for", async () => {
    const template = `
      <kpn-list-item v-for="item in items" :key="item.id">
        <kpn-title id="title" :foo="bar">Hello world!</kpn-title>
      </kpn-list-item>
    `;

    const string = await autoLitWrapperOld(template, "LitWrapper", filter);
    expect(string).toMatchSnapshot();
  });

  test("v-if and v-else", async () => {
    const template = `
      <kpn-container v-if="true">
        <kpn-title id="title" :foo="bar">Hello world!</kpn-title>
      </kpn-container>
      <kpn-container v-else>
        <kpn-title id="title" :foo="bar">Hello world 2!</kpn-title>
      </kpn-container>
    `;

    const string = await autoLitWrapperOld(template, "LitWrapper", filter);
    expect(string).toMatchSnapshot();
  });
});
