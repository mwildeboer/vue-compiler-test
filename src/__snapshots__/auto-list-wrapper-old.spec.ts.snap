// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`autoLitWrapper > complex 1`] = `
"
    <LitWrapper v-if="true"><kpn-title id="title" :foo=""bar"> 1 ? 2 : 1" bind:test="test">
        <span>Hello world!</span>
        <img src="src">
    </kpn-title></LitWrapper>
    "
`;

exports[`autoLitWrapper > for 1`] = `
"
      <LitWrapper v-for="item in items" :key="item.id"><kpn-list-item>
        <LitWrapper><kpn-title id="title" :foo="bar">Hello world!</kpn-title></LitWrapper>
      </kpn-list-item></LitWrapper>
    "
`;

exports[`autoLitWrapper > simple 1`] = `"<LitWrapper><kpn-title id="title" :foo="bar">Hello world!</kpn-title></LitWrapper>"`;

exports[`autoLitWrapper > v-if and v-else 1`] = `
"
      <LitWrapper v-if="true"><kpn-container>
        <LitWrapper><kpn-title id="title" :foo="bar">Hello world!</kpn-title></LitWrapper>
      </kpn-container></LitWrapper>
      <LitWrapper v-else=""><kpn-container>
        <LitWrapper><kpn-title id="title" :foo="bar">Hello world 2!</kpn-title></LitWrapper>
      </kpn-container></LitWrapper>
    "
`;
